lib/build.ts

```diff
@@ -328,32 +333,19 @@ async function checkAllowDockerPlatformHandling(
 		return true;
 	} else if (imagesWithPlatformSupport.length > 0) {
 		// Mixed images
-		console.log(
-			'Warning: Dockerfile contains a mixes images that require a platform selection with ones that do not support it.\n' +
-				'The following images do not support platform selection:\n-' +
-				imagesWithoutPlatformSupport.join('\n-') +
-				'\n\n' +
-				'The following images require platform selection:\n-' +
-				imagesWithPlatformSupport.join('\n-') +
-				'\n\n' +
-				'To avoid a possible build error, the CLI has disabled platform selection. ' +
-				'As a result, the architecture of the machine where the Docker Engine ' +
-				'is running will be used to select the platform when pulling upstream images ' +
-				'which may result in "exec format error" at runtime. To avoid this warning, ' +
-				'update or replace the images that do not support platform selection.',
-		);
-	} else {
-		// Only images without platform support
-		console.log(
-			'Warning: Dockerfile contains a images do not support platform selection.\n-' +
-				imagesWithoutPlatformSupport.join('\n-') +
-				'\n\n' +
-				'To avoid a possible build error, the CLI has disabled platform selection. ' +
-				'As a result, the architecture of the machine where the Docker Engine ' +
-				'is running will be used to select the platform when pulling upstream images ' +
-				'which may result in "exec format error" at runtime. To avoid this warning, ' +
-				'update or replace the images that do not support platform selection.',
-		);
+		warn(`\
+Service '${task.serviceName}':
+  Multi-stage Dockerfile found with a mix of base images that require
+  CPU architecture selection and base images that do not support it.
+  The following base images do not support CPU architecture selection:
+  - ${imagesWithoutPlatformSupport.join('\n  - ')}
+  The following base images require CPU architecture selection:
+  - ${imagesWithPlatformSupport.join('\n  - ')}
+  As a result, the CPU architecture of the machine where the Docker Engine
+  is running will be used to select base images that require architecture
+  selection.  This may result in incorrect architecture selection and
+  "exec format error" at runtime. To avoid this warning, update or replace
+  base images that do not support architecture selection.`);
```

Comment:
Here's how this ends up looking with the CLI -- the lines prefixed with `[Warn]`:
```
$ ./bin/balena build t -A armv7hf -d raspberrypi3 --nocache --pull --debug
...
[Info]    Building for armv7hf/raspberrypi3
[Info]    Docker Desktop detected (daemon architecture: "x86_64")
[Info]      Docker itself will determine and enable architecture emulation if required,
[Info]      without balena-cli intervention and regardless of the --emulated option.
[Debug]   Found build tasks:
[Debug]       main: build [.]
[Debug]   Resolving services with [raspberrypi3|armv7hf]
[Debug]   Found project types:
[Debug]       main: Standard Dockerfile
[Debug]   Prepared tasks; building...
[Warn]    Service 'foo':
[Warn]      Multi-stage Dockerfile found with a mix of base images that require
[Warn]      CPU architecture selection and base images that do not support it.
[Warn]      The following base images do not support CPU architecture selection:
[Warn]      - balenalib/raspberrypi3-node:14-buster-run
[Warn]      The following base images require CPU architecture selection:
[Warn]      - debian:10
[Warn]      - busybox
[Warn]      As a result, the CPU architecture of the machine where the Docker Engine
[Warn]      is running will be used to select base images that require architecture
[Warn]      selection.  This may result in incorrect architecture selection and
[Warn]      "exec format error" at runtime. To avoid this warning, update or replace
[Warn]      base images that do not support architecture selection.
[Build]   main Step 1/5 : FROM balenalib/raspberrypi3-node:14-buster-run as base
[Build]   main  ---> e074e38a89ba
```

Reply:
Regarding the advice:
```
To avoid this warning, update or replace base images that do not support architecture selection.
```
Ideally the latest balenalib base images would qualify as base images that support architecture selection, otherwise we are effectively telling balena users to replace balenalib base images with non-balenalib base images.

I have gathered that `balenalib/raspberrypi3-node:latest` is treated by `balena-multibuild` as not supporting platform selection.

I see that [a discussion started in flowdock](https://www.flowdock.com/app/rulemotion/i-cli/threads/RuSu1KiWOn62xaGy7O2sn8m8BUc) with @nghiant2710, who appears to claim that the latest balenalib base images correctly declare their platform or were generated as they should. Well, either `balena-multibuild` is not correctly detecting this, or balenalib base images are still missing something...

Reply:
I have just tested `FROM --platform` with this multi-stage `Dockerfile.template`:
```Dockerfile
FROM balenalib/%%BALENA_MACHINE_NAME%%-node as base
FROM --platform=linux/arm/v7 busybox as base2
FROM --platform=linux/arm/v7 debian:10
COPY --from=base /usr/local/bin/node /tmp/node
COPY --from=base2 /bin/busybox /tmp/busybox
CMD ["/bin/bash"]
```

And `balena build -A armv7hf -d raspberrypi3`, thus meaning to build for ARM, but running on an Intel laptop. The CLI prints the mixed-images warning as expected. The expected, but still interesting observation is that:

* Without `FROM --platform=linux/arm/v7` (no FROM-line platform selection), the `busybox` and `debian` images are selected for Intel, as the warning warns. The templated `rasbperrypi3-node` base image is for ARM, of course, as it is a single-arch image.
* With `FROM --platform=linux/arm/v7`, the `busybox` and `debian` images are selected for ARM v7, also as expected.

Therefore, instead of advising `To avoid this warning, update or replace base images that do not support architecture selection` (especially if this was not yet feasible with `balenalib`), we could instead advise the possibility of using `FROM --platform`, if users needed to mix base images. Perhaps:

```ts
  `As a result, the CPU architecture of the machine where the Docker Engine
  is running will be used by default to select base images that require
  architecture selection. This may result in incorrect architecture selection
  and "exec format error" at runtime. It is usually possible to override the
  architecture in the FROM line with e.g. "FROM --platform=linux/arm/v7",
  or by adding the sha256 digest of the image for a specific architecture
  with e.g. "FROM debian@sha256:094f5..."`);
```

Also:
* As we know the target architecture (`-A armv7hf -d raspberrypi3`),
* And we know we cannot add a common `platform` option to the Docker API (as Docker Engine will throw an error, and the reason why we print the mixed-image warning),
* And we know which images require architecture selection (we are printing the list in the warning message),
* And we know that `FROM --platform=linux/arm/v7` would solve the problem (I'm planning to add this advice to the warning message),
* Then (in another PR) `balena-multibuild` could automatically and on-the-fly add `FROM --platform=linux/arm/v7` where needed (in the build stream only, not saved to disk).

Reply:
see:  https://github.com/balena-io-modules/balena-multibuild/issues/97